### ISSUE TITLE
`amber generate auth` - a user cannot be updated with also updated `password`

### ISSUE THREAD

**Issue by marksiemers** [amberframework/amber#392]
### Description

The auth generator creates this validation:
```
  validate :password, "is to short", -> (user : User) do
    (user.password != nil) && (user.password.not_nil!.size >= 8)
  end
```
Which requires that on update (any `save`) the validation will fail and the user will not be saved.

The design of the model needs to change to allow for updates to other fields without requiring an update to `password`

### Steps to Reproduce

1. `amber new blog && cd blog && shards install`
2. `amber g auth User is_admin:bool` 
3. `amber db create migrate seed`
4. `amber exec 'u = User.first'`
5. `amber exec 'u = User.first; u.not_nil!.is_admin = true; u.not_nil!.save; u.not_nil!.errors;'`

**Expected behavior:** [What you expect to happen]

The user to save without requiring the password field. In the case that this update was made from the web, the call to `authenticate` could and should happen.

**Actual behavior:** [What actually happens]

```shell
$ amber exec 'u = User.first'
#<User:0x10b3bc7e0 @errors=[], @password=nil, @id=1, @is_admin=true, @email="admin@example.com", @encrypted_password="$2a$10$1BnfLjGC.7653y05UptrnOHqFr4NL0zWP6WXSawTlZuclKCt5Qp0i", @created_at=2017-11-14 10:59:22 UTC, @updated_at=2017-11-14 11:09:50 UTC>

$ amber exec 'u = User.first; u.not_nil!.is_admin = true; u.not_nil!.save; u.not_nil!.errors;'
Save Exception: Nil assertion failed
[#<Granite::ORM::Error:0x1077115c0 @field=:base, @message="Nil assertion failed">]
```

**Reproduces how often:** [What percentage of the time does it reproduce?]

100%

### Versions

amber master
crystal 0.23.1

### Additional Information

This may merit discussion, as often an update to a user should require a password - but it should be used to authorize, rather than update.

**Comment by faustinoaq** [comment]
We should avoid using `.not_nil` everywhere:

An alternative is using `if`:

```crystal
  validate :password, "is to short", -> (user : User) do
    if password = user.password
      password.size >= 8
    end
  end
```

```sh
$ amber exec 'if u = User.first; u.is_admin = true; u.save; u.errors; end'
```

**Comment by marksiemers** [comment]
Indeed, I couldn't figure out the one-liner syntax, tried `if u then ... end` and `if (u) statement`

**Comment by faustinoaq** [comment]
Well `if .. then` not exist in crystal because `then` keyword isn't available for if statement

On `amber exec` I would recommend to use heredoc for multiple lines:

So it looks more clear:

```crystal
$ amber exec <<CODE
> if u = User.first
> u.is_admin = true
> u.save
> u.errors
> end
> CODE
# output ...
```

**Comment by marksiemers** [comment]
That's a good idea.

I went with post conditions:
```
$ amber exec 'a = Article.find(1); (a.archived = false) if a; a.save if a;'
```

**Comment by marksiemers** [comment]
@drujensen - Any thoughts on this issue?

**Comment by drujensen** [comment]
yeah, we need to change the validation a bit.  We should break into two different validations, existence and length.

1. To check if a password has been set, we need to verify the `encrypted_password` exists or the `password` is being set.

2. To check the length, only validate the length when a password field is set.

**Comment by marksiemers** [comment]
So, I haven't been able to update the controllers/views to test this fully, but here is a shot at redesigning the class generated by `amber g auth User`

@drujensen - WDYT?
```crystal
require "granite_orm/adapter/pg"
require "crypto/bcrypt/password"

class User < Granite::ORM::Base
  include Crypto
  adapter pg
  primary id : Int64
  field email : String
  field encrypted_password : String
  timestamps

  validate :email, "is required", -> (user : User) do
    (email = user.email) ? !email.empty? : false
  end

  validate :password, "is too short", -> (user : User) do
    user.password_changed? ? user.valid_password_size? : true
  end

  def password=(password)
    @new_password = password
    @encrypted_password = Bcrypt::Password.create(password, cost: 10).to_s
  end

  def password
    @password ||= (enc = encrypted_password) ? Bcrypt::Password.new(enc) : nil
  end

  def password_changed?
    new_password ? true : false
  end

  def valid_password_size?
    (pass = new_password) ? pass.size >= 8 : false
  end

  def authenticate(password : String)
    (bcrypt_pass = self.password) ? bcrypt_pass == password : false
  end

  private getter new_password : String?
end
```

**Comment by marksiemers** [comment]
I have a working app example here: https://github.com/marksiemers/auth-test

It could use some polish, but the user model, controller, and views are working for registration, login, and updating your profile.

**Comment by drujensen** [comment]
Thanks @marksiemers Do you want to open a PR?  we can collaborate on it.

**Comment by marksiemers** [comment]
Sure thing, I may not be able to move the code over until tomorrow.

**Comment by marksiemers** [comment]
#407

**Comment by drujensen** [comment]
This issue has been addressed by #407
